fix lost tsCache updates

tbg opened the pull request:
previously, concurrent reads during the execution of the
splitTrigger could add updates to the timestamp cache
on the old range after it had already been copied to
the new range. With this change, reads happen atomically
with respect to commit triggers.

fixes #3148.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3175)

<!-- Reviewable:end -->

tamird commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[storage/replica_command.go, line 1400 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175#-K3VrrEyu735tW5qf0wP:-K3VrrEyu735tW5qf0wQ:979220883) ([raw file](https://github.com/cockroachdb/cockroach/blob/7661963d1c65b31dda9625a50a899adbca700777/storage/replica_command.go#L1400)):_
this seems weird; if this lock isn't necessary, remove it.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175)_
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[storage/replica_command.go, line 1400 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175#-K3VrrEyu735tW5qf0wP:-K3VtbNvdtVg6S_5MmaZ:-1758509903) ([raw file](https://github.com/cockroachdb/cockroach/blob/7661963d1c65b31dda9625a50a899adbca700777/storage/replica_command.go#L1400)):_
It's not necessary here because I know (or at least am reasonably certain) that nobody can concurrently access it with `readMu` locked **at the time of writing**, but tsCache is one of the fields protected by the replica lock and so I should grab it regardless. I can remove the comment if you prefer that.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175)_
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

_[storage/replica_command.go, line 1400 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175#-K3VrrEyu735tW5qf0wP:-K3VuEVVFJRWU9yPLkOd:-1683889609) ([raw file](https://github.com/cockroachdb/cockroach/blob/7661963d1c65b31dda9625a50a899adbca700777/storage/replica_command.go#L1400)):_
Fine as is.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175)_
<!-- Sent from Reviewable.io -->

bdarnell commented on storage/replica.go:
```diff
@@ -163,6 +163,7 @@ type Replica struct {
 	// proposeRaftCommandFn can be set to mock out the propose operation.
 	proposeRaftCommandFn func(cmdIDKey, roachpb.RaftCommand) <-chan error
 
+	readMu       sync.RWMutex    // Held during read-only commands
```
`readMu` sounds confusingly similar to the `RLock` of the embedded `RWMutex`. Maybe call it `readOnlyCmdMu`?

I'd add more detail to the comment: "Held in read mode during read-only commands. Held in exclusive mode to prevent read-only commands from executing. Acquired before the embedded `RWMutex`"

tschottdorf replied:
done.

bdarnell commented on storage/replica_command.go:
```diff
@@ -473,6 +473,13 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, h ro
+			// This makes sure that no reads are happening in parallel;
+			// see #3148.
+			r.readMu.Lock()
+			batch.Defer(r.readMu.Unlock)
```
Some commit triggers also use batch.Defer; it might be worth a comment that these will run outside the scope of `readMu`.

tschottdorf replied:
But they'll run inside of the scope, won't they? After all, this is the first defer added.

bdarnell replied:
The commit triggers themselves run inside the scope of the lock; the deferred portions are added later so they run outside the lock.

bdarnell commented:
Can you add a test that deliberately races a split with a read-only command? It doesn't have to consistently trigger the issue but it would be good to have a more reliable and explicit test than the rare failure of sql.TestConcurrentIncrements. (which only triggered this issue because of the non-obvious fact that table creation leads to an asynchronous split)

tschottdorf commented:
I'll see what I can do about the test.

tbg committed:
fix lost tsCache updates

previously, concurrent reads during the execution of the
splitTrigger could add updates to the timestamp cache
on the old range after it had already been copied to
the new range. With this change, reads happen atomically
with respect to commit triggers.

fixes #3148.

tschottdorf commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

_[storage/replica_command.go, line 481 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175#-K3VpY7L6OFWIBPCvMxb-r1-481:-K3XGOUKcYL6806tsD7y:-922118152) ([raw file](https://github.com/cockroachdb/cockroach/blob/7661963d1c65b31dda9625a50a899adbca700777/storage/replica_command.go#L481)):_
I must be thick but to me it still looks like the first `batch.Defer`ed action is the unlock, so that should be the very last thing called (i.e. all the other `batch.Defer`ed functions should be under the lock).

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175)_
<!-- Sent from Reviewable.io -->

tbg committed:
add regression test for #3148

tbg committed:
log coord-side error in traces

tbg committed:
clarify/harden locking around SplitTriggers

tschottdorf commented:
PTAL @bdarnell. I added a test that's flaky as hell before this fix, and also took into account the fact that the `batch.Defer()` would not run if the commit trigger returned an error, deadlocking the range.

tamird commented:
Reviewed 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[storage/client_split_test.go, line 790 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175#-K3XXXsAC0G7rJ3ygkOc:-K3XXXsAC0G7rJ3ygkOd:677817405) ([raw file](https://github.com/cockroachdb/cockroach/blob/9117dcaba7e16fbe45e704020356860f488ff9f8/storage/client_split_test.go#L790)):_
lets

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175)_
<!-- Sent from Reviewable.io -->

tschottdorf commented:
`TestRaftRemoveRace` fails very reliably on this branch. The new lock must've changed timings in a way that make the errors much more likely. See #3178 #3010.

bdarnell commented:
LGTM

---

Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4.
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

_[storage/replica_command.go, line 481 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175#-K3VpY7L6OFWIBPCvMxb-r1-481:-K3b0JRDRRBmNR_wMlW1:-872991641) ([raw file](https://github.com/cockroachdb/cockroach/blob/7661963d1c65b31dda9625a50a899adbca700777/storage/replica_command.go#L481)):_
No, it's me being thick; I looked at `batch.Defer` and didn't notice we reversed the list when we called the functions.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3175#-:-K3b24-WQBJyjCCmBAV2:676834233)_
<!-- Sent from Reviewable.io -->
